src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Package/Update/PackageUpdateCommandRunner.cs

```diff
@@ -278,194 +275,93 @@ bool PackageHasVulnerability(string packageId, NuGetVersion version, IReadOnlyLi
+                if (latestVersion is null)
                 {
-                    if (allPackageIds.Add(dependency.Name))
-                    {
-                        allPackages.Add(new Package
-                        {
-                            Id = dependency.Name,
-                            VersionRange = dependency.LibraryRange.VersionRange
-                        });
-                    }
+                    logger.LogMinimal(Messages.Error_NoVersionsAvailable(package.Id), ConsoleColor.Red);
+                    hasErrors = true;
+                    continue;
+                }
+
+                upgradeVersion = new VersionRange(latestVersion);
+                if (upgradeVersion == existingVersion)
+                {
+                    logger.LogMinimal(Messages.Warning_AlreadyHighestVersion(package.Id, latestVersion.OriginalVersion, project.FilePath), ConsoleColor.Yellow);
+                    continue;
                 }
             }
+
+            var packageToUpdate = new PackageToUpdate
+            {
+                Id = package.Id,
+                CurrentVersion = existingVersion,
+                NewVersion = upgradeVersion
+            };
+            packagesToUpdate.Add(new PackageUpdateResult
+            {
+                Package = packageToUpdate,
+                TargetFrameworkAliases = packageTfms
+            });
         }
 
-        return allPackages;
+        return hasErrors ? [] : packagesToUpdate;
     }
 
-    private static async Task<PackageUpdateResult?> GetSinglePackageToUpdateAsync(
-        Package package,
-        PackageSpec project,
-        IVersionChooser versionChooser,
-        ISettings settings,
-        ILoggerWithColor logger,
-        bool suppressWarnings,
-        CancellationToken cancellationToken)
+    /// <summary>Gets the package's referenced version range and TFMs which it's referenced in.</summary>
+    /// <param name="packageId">The package identifier to find.</param>
+    /// <param name="project">The project's restore inputs.</param>
+    /// <param name="logger">The logger to output errors.</param>
+    /// <returns>
+    /// <para>When the package is found and no problems occur, it returns the requested <see cref="VersionRange"/>
+    /// and list of target framework aliases that the package is used in.</para>
+    /// <para>If the <see cref="VersionRange"/> is null and the target framework list contains at least one value,
+    /// then there /// was an error, which the method will have logged. Note that the target framework list
+    /// may not be complete.</para>
+    /// <para>When the version range is null and the target framework list is empty, it means that the project
+    /// does not reference the package (and no error was logged).</para>
+    /// </returns>
+    private static (VersionRange? version, List<string> targetFrameworks)
```

Comment:
Longer term (looking now couldn't find an existing API), I think this API probably belongs in PackageSpecOperations or something similar.

Reply:
Yeah makes sense. 

I think this API has usages in dotnet add/update commands, as well as VS installation code and probably solver.